enterprise/backend/src/metabase_enterprise/serialization/api.clj

```diff
@@ -0,0 +1,237 @@
+(ns metabase-enterprise.serialization.api
+  (:require
+   [clojure.java.io :as io]
+   [compojure.core :refer [POST]]
+   [java-time.api :as t]
+   [metabase-enterprise.serialization.v2.extract :as extract]
+   [metabase-enterprise.serialization.v2.ingest :as v2.ingest]
+   [metabase-enterprise.serialization.v2.load :as v2.load]
+   [metabase-enterprise.serialization.v2.storage :as storage]
+   [metabase.api.common :as api]
+   [metabase.api.routes.common :refer [+auth]]
+   [metabase.logger :as logger]
+   [metabase.models.serialization :as serdes]
+   [metabase.public-settings :as public-settings]
+   [metabase.util :as u]
+   [metabase.util.date-2 :as u.date]
+   [metabase.util.log :as log]
+   [metabase.util.malli.schema :as ms]
+   [ring.core.protocols :as ring.protocols])
+  (:import
+   (java.io File ByteArrayOutputStream)
+   (java.lang AutoCloseable)
+   (org.apache.commons.compress.archivers.tar TarArchiveEntry TarArchiveInputStream TarArchiveOutputStream)
+   (org.apache.commons.compress.compressors.gzip GzipCompressorInputStream GzipCompressorOutputStream GzipParameters)
+   (org.apache.logging.log4j Level)
+   (org.apache.logging.log4j.core.appender AbstractAppender FileAppender)
+   (org.apache.logging.log4j.core.config AbstractConfiguration LoggerConfig)))
+
+(set! *warn-on-reflection* true)
+
+(def ^:dynamic *in-tests* "Set when executed in tests to prevent async removal of files" false)
+
+;;; Request callbacks
+
+(defn- ba-copy [f]
+  (let [baos (ByteArrayOutputStream.)]
+    (io/copy f baos)
+    (.toByteArray baos)))
+
+(defn- on-response! [data callback]
+  (reify
+    ;; Real HTTP requests and mt/user-real-request go here
+    ring.protocols/StreamableResponseBody
+    (write-body-to-stream [_ response out]
+      (ring.protocols/write-body-to-stream data response out)
+      (if *in-tests*
```

Comment:
```suggestion
      (if *in-tests*
```

binding this in tests shouldn't work because `mt/user-real-request` runs in its own thread.

it should only works with `mt/user-http-request`.

Reply:
Hahaha, right! Well, it doesn't make sense for `user-http-request` since tests need no use of `future`. I wonder if passing bindings to a thread inside `user-real-request` will break anything?.. I really don't want to introduce a potential flake here. :)

Do you know why it's in a separate thread, is there a reason for this?

Reply:
> Well, it doesn't make sense for user-http-request since tests need no use of future

I thought you wanted to make sure the call back is called synchronously? 

> Do you know why it's in a separate thread, is there a reason for this?

I'm not sure I understand the question. anw we're making a real HTTP request with `user-real-request`, so the thread that handle the request is different from the thread we make the request.

Reply:
Well, there is no point in having `(future)` inside `make-input-stream`, since this _only_ used for tests, not for the real service. So right now, if `user-real-request` goes through the network (I somehow imagined it still went through some inner call) this potentially introduces a flake with future in the `write-body-to-stream`. Not sure if it's worth it then...

Reply:
yea, I think we can remove `future` for `make-input-stream`, but we definitely should use `future` in `write-body-to-stream`

Reply:
bump

Reply:
ah right, I need to remove `*in-tests*`. Now the only problem is that I'm not sure how to control that `future` is executed in time and this test does not become a flake. Any thoughts on this?..